lib/groups.js

```diff
@@ -118,33 +276,182 @@ module.exports.createGroup = function (groupName, assessmentId, userId, authnUse
+      callback(null, false, null, invalidGroupName, groupConfig);
     });
   }
 };
 
-module.exports.leaveGroup = function (assessmentId, userId, authnUserId, callback) {
-  const params = {
-    assessment_id: assessmentId,
-    user_id: userId,
-    authn_user_id: authnUserId,
-  };
-  sqldb.query(sql.leave_group, params, function (err) {
-    if (ERR(err, callback)) return;
-    callback();
+/**
+ * @param {string} assessmentId
+ * @param {string} userId
+ * @param {string} authnUserId
+ */
+module.exports.leaveGroup = async function (assessmentId, userId, authnUserId) {
+  await sqldb.runInTransactionAsync(async () => {
+    const groupId = await module.exports.getGroupId(assessmentId, userId);
+    if (groupId === null) {
+      throw new Error(
+        "Couldn't access the user's group ID with the provided assessment and user IDs"
+      );
+    }
+    const groupConfig = await module.exports.getGroupConfig(assessmentId);
+
+    if (groupConfig.has_roles) {
+      const groupInfo = await module.exports.getGroupInfo(groupId, groupConfig);
+
+      // Reassign roles if there is more than 1 user
+      const currentSize = groupInfo.groupMembers.length;
+      if (currentSize > 1) {
+        // Get leaving user's required roles
+        const requiredRoles = groupInfo.rolesInfo.groupRoles.filter(
+          (role) => (role.minimum ?? 0) > 0
+        );
+        const groupRoleAssignments = Object.values(groupInfo.rolesInfo.roleAssignments).flat();
+        const requiredGroupRoleIds = groupRoleAssignments
+          .filter(({ user_id }) => user_id === userId.toString())
+          .map(({ group_role_id }) => group_role_id)
+          .filter((roleId) => requiredRoles.some((role) => role.id === roleId));
+
+        if (requiredGroupRoleIds.length > 0) {
+          // Get group user to group role assignments, excluding the leaving user
+          const groupRoleAssignments = Object.values(groupInfo.rolesInfo.roleAssignments)
+            .flat()
+            .filter(({ user_id }) => user_id !== userId.toString());
+
+          // Replace non-required group role assignments with required roles from leaving user
+          const groupRoleAssignmentUpdates = groupRoleAssignments.map(
+            ({ user_id, group_role_id }) => {
+              const notRequired = !requiredRoles.some((role) => role.id === group_role_id);
+              if (notRequired && requiredGroupRoleIds.length > 0) {
+                return {
+                  user_id,
+                  group_role_id: requiredGroupRoleIds.pop(),
+                  group_id: groupId,
+                };
+              } else {
+                return { user_id, group_role_id, group_id: groupId };
+              }
+            }
+          );
+
+          // Assign remaining required roles to another group user
+          const assigneeUser = groupInfo.groupMembers.find((m) => m.user_id !== userId.toString());
+          if (assigneeUser === undefined) {
```

Comment:
`assigneeUser` should never be `undefined` here, because we're checking above that `currentSize > 1. Is throwing an error like this the cleanest way to avoid type checking errors? Ditto for other errors thrown in this file for similar `null`/`undefined` checks.

Reply:
Yep - throwing an error is the way to go!